<a href="https://colab.research.google.com/github/RSkrep/deep-learning-challenge/blob/main/AlphabetSoupCharity_Optimization2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Preprocessing

In [2]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

#  Import and read the charity_data.csv.
import pandas as pd
application_df = pd.read_csv("https://static.bc-edx.com/data/dl-1-2/m21/lms/starter/charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [6]:
application_df['NAME'].value_counts()

NAME
PARENT BOOSTER USA INC                                                  1260
TOPS CLUB INC                                                            765
UNITED STATES BOWLING CONGRESS INC                                       700
WASHINGTON STATE UNIVERSITY                                              492
AMATEUR ATHLETIC UNION OF THE UNITED STATES INC                          408
                                                                        ... 
ST LOUIS SLAM WOMENS FOOTBALL                                              1
AIESEC ALUMNI IBEROAMERICA CORP                                            1
WEALLBLEEDRED ORG INC                                                      1
AMERICAN SOCIETY FOR STANDARDS IN MEDIUMSHIP & PSYCHICAL INVESTIGATI       1
WATERHOUSE CHARITABLE TR                                                   1
Name: count, Length: 19568, dtype: int64

In [7]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'. Keep Name now
#  YOUR CODE GOES HERE
application_df_new = application_df.drop (['EIN'], axis=1)

In [8]:
# Determine the number of unique values in each column.
#  YOUR CODE GOES HERE
unique_values = application_df_new.nunique()
print(unique_values)


NAME                      19568
APPLICATION_TYPE             17
AFFILIATION                   6
CLASSIFICATION               71
USE_CASE                      5
ORGANIZATION                  4
STATUS                        2
INCOME_AMT                    9
SPECIAL_CONSIDERATIONS        2
ASK_AMT                    8747
IS_SUCCESSFUL                 2
dtype: int64


In [9]:
# Look at APPLICATION_TYPE value counts for binning
#  YOUR CODE GOES HERE
application_type_counts = application_df_new['APPLICATION_TYPE'].value_counts()
print(application_type_counts)


APPLICATION_TYPE
T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: count, dtype: int64


In [10]:
# Choose a cutoff value and create a list of names to be replaced
# use the variable name `names_to_replace`
#  YOUR CODE GOES HERE
cutoff_value = 5
# Create a list of Names to be replaced
application_types_to_replace = application_df_new['NAME'].value_counts()[application_df_new['NAME'].value_counts() < cutoff_value].index.tolist()
# Replace in dataframe
for app in application_types_to_replace:
    application_df_new['NAME'] = application_df_new['NAME'].replace(app,"Other")

# Check to make sure binning was successful
application_df_new['NAME'].value_counts()

NAME
Other                                                                19803
PARENT BOOSTER USA INC                                                1260
TOPS CLUB INC                                                          765
UNITED STATES BOWLING CONGRESS INC                                     700
WASHINGTON STATE UNIVERSITY                                            492
                                                                     ...  
GUILDS OF THE SANTA FE OPERA INC                                         5
FAMILY MOTOR COACH ASSOCIATION                                           5
JAMESTOWN SOCIETY                                                        5
MOST WORSHIPFUL PRINCE HALL GRAND LODGE OF FREE & ACCEPTED MASONS        5
PTA HAWAII CONGRESS                                                      5
Name: count, Length: 403, dtype: int64

In [11]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
#  YOUR CODE GOES HERE
cutoff_value = 100
# Create a list of application types to be replaced
application_types_to_replace = application_df_new['APPLICATION_TYPE'].value_counts()[application_df_new['APPLICATION_TYPE'].value_counts() < cutoff_value].index.tolist()
# Replace in dataframe
for app in application_types_to_replace:
    application_df_new['APPLICATION_TYPE'] = application_df_new['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure binning was successful
application_df_new['APPLICATION_TYPE'].value_counts()

APPLICATION_TYPE
T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
T9         156
Other      120
Name: count, dtype: int64

In [12]:
# Look at CLASSIFICATION value counts for binning
#  YOUR CODE GOES HERE
classification_counts = application_df_new['CLASSIFICATION'].value_counts()
print(classification_counts)


CLASSIFICATION
C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: count, Length: 71, dtype: int64


In [13]:
# You may find it helpful to look at CLASSIFICATION value counts >1
#  YOUR CODE GOES HERE
classification_counts_greater = classification_counts[classification_counts >1]
print(classification_counts_greater)

CLASSIFICATION
C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
C7120       18
C1500       16
C1800       15
C6000       15
C1250       14
C8200       11
C1238       10
C1278       10
C1235        9
C1237        9
C7210        7
C2400        6
C1720        6
C4100        6
C1257        5
C1600        5
C1260        3
C2710        3
C0           3
C3200        2
C1234        2
C1246        2
C1267        2
C1256        2
Name: count, dtype: int64


In [14]:
# Choose a cutoff value and create a list of classifications to be replaced
# use the variable name `classifications_to_replace`
#  YOUR CODE GOES HERE
cutoff_value = 100
classifications_to_replace = application_df_new['CLASSIFICATION'].value_counts()[application_df_new['CLASSIFICATION'].value_counts() < cutoff_value].index.tolist()
# Replace in dataframe
for cls in classifications_to_replace:
    application_df_new['CLASSIFICATION'] = application_df_new['CLASSIFICATION'].replace(cls,"Other")

# Check to make sure binning was successful
application_df_new['CLASSIFICATION'].value_counts()

CLASSIFICATION
C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
Other      669
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
Name: count, dtype: int64

In [15]:
# Convert categorical data to numeric with `pd.get_dummies`
#  YOUR CODE GOES HERE
data_encoded = pd.get_dummies(application_df_new)

data_encoded

STATUS   ASK_AMT  IS_SUCCESSFUL  NAME_AACE INTERNATIONAL  \
0           1      5000              1                    False   
1           1    108590              1                    False   
2           1      5000              0                    False   
3           1      6692              1                    False   
4           1    142590              1                    False   
...       ...       ...            ...                      ...   
34294       1      5000              0                    False   
34295       1      5000              0                    False   
34296       1      5000              0                    False   
34297       1      5000              1                    False   
34298       1  36500179              0                    False   

       NAME_ACE MENTOR PROGRAM OF AMERICA INC  NAME_ACTS MINISTRY  \
0                                       False               False   
1                                       False               False   
2                                       False               False   
3                                       False               False   
4                                       False               False   
...                                       ...                 ...   
34294                                   False               False   
34295                                   False               False   
34296                                   False               False   
34297                                   False               False   
34298                                   False               False   

       NAME_ACTS MISSIONS  \
0                   False   
1                   False   
2                   False   
3                   False   
4                   False   
...                   ...   
34294               False   
34295               False   
34296               False   
34297               False   
34298               False   

       NAME_AFRICAN-AMERICAN POSTAL LEAGUE UNITED FOR SUCCESS A-PLUS  \
0                                                  False               
1                                                  False               
2                                                  False               
3                                                  False               
4                                                  False               
...                                                  ...               
34294                                              False               
34295                                              False               
34296                                              False               
34297                                              False               
34298                                              False               

       NAME_AIR FORCE ASSOCIATION  NAME_ALABAMA FEDERATION OF WOMENS CLUBS  \
0                           False                                    False   
1                           False                                    False   
2                           False                                    False   
3                           False                                    False   
4                           False                                    False   
...                           ...                                      ...   
34294                       False                                    False   
34295                       False                                    False   
34296                       False                                    False   
34297                       False                                    False   
34298                       False                                    False   

       ...  INCOME_AMT_1-9999  INCOME_AMT_10000-24999  \
0      ...              False                   False   
1      ...               True                   False   
2      ...              False                   Fal

In [16]:
# Split our preprocessed data into our features and target arrays
#  YOUR CODE GOES HERE
y= data_encoded['IS_SUCCESSFUL'].values
X= data_encoded.drop(['IS_SUCCESSFUL'], axis='columns').values
# Split the preprocessed data into a training and testing dataset
#  YOUR CODE GOES HERE
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=78)

In [17]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Compile, Train and Evaluate the Model

In [18]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
#  YOUR CODE GOES HER
nn = tf.keras.models.Sequential()
number_input_features = len(X_train[0])
hidden_nodes_layer1 =  100
hidden_nodes_layer2 = 30
hidden_nodes_layer3 = 10
# First hidden layer
#  YOUR CODE GOES HERE
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer1, activation="relu", input_dim=number_input_features))

# Second hidden layer
#  YOUR CODE GOES HERE
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer2, activation="relu"))
# Third Hidden Layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer3, activation="relu"))

# Output layer
#  YOUR CODE GOES HERE
nn.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 100)               45400     
                                                                 
 dense_1 (Dense)             (None, 30)                3030      
                                                                 
 dense_2 (Dense)             (None, 10)                310       
                                                                 
 dense_3 (Dense)             (None, 1)                 11        
                                                                 
Total params: 48751 (190.43 KB)
Trainable params: 48751 (190.43 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [19]:
# Compile the model
#  YOUR CODE GOES HERE
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])


In [20]:
# Train the model
#  YOUR CODE GOES HERE
fit_model = nn.fit(X_train_scaled, y_train, epochs=100)

Epoch 1/100
804/804 [==============================] - 7s 7ms/step - loss: 0.4712 - accuracy: 0.7783
Epoch 2/100
804/804 [==============================] - 5s 7ms/step - loss: 0.4289 - accuracy: 0.7942
Epoch 3/100
804/804 [==============================] - 3s 3ms/step - loss: 0.4246 - accuracy: 0.7956
Epoch 4/100
804/804 [==============================] - 2s 3ms/step - loss: 0.4227 - accuracy: 0.7973
Epoch 5/100
804/804 [==============================] - 2s 2ms/step - loss: 0.4205 - accuracy: 0.7984
Epoch 6/100
804/804 [==============================] - 2s 3ms/step - loss: 0.4197 - accuracy: 0.7982
Epoch 7/100
804/804 [==============================] - 2s 2ms/step - loss: 0.4184 - accuracy: 0.7985
Epoch 8/100
804/804 [==============================] - 3s 4ms/step - loss: 0.4178 - accuracy: 0.7997
Epoch 9/100
804/804 [==============================] - 2s 3ms/step - loss: 0.4170 - accuracy: 0.8009
Epoch 10/100
804/804 [==============================] - 2s 2ms/step - loss: 0.4163 - accura

In [21]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - loss: 0.4565 - accuracy: 0.7903 - 525ms/epoch - 2ms/step
Loss: 0.45652857422828674, Accuracy: 0.7903206944465637


In [22]:
# Export our model to HDF5 file
#  YOUR CODE GOES HERE
nn.save('AlphabetSoupCharity_Optimization2.h5')

/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


Overview of the Analysis:

The purpose of this analysis was to develop a machine learning algorithm to predict the success of charities. Utilizing deep learning techniques, three machine learning models were created, with the final model achieving an accuracy score greater than 75%. Each iteration of the model was adjusted slightly to optimize its performance and provide the best predictive tool.

Results:

Data Preprocessing:

Target Variable(s):

The target variable for each model was the Is_SUCCESSFUL column, as it provided crucial information about the success of past charities and was the primary focus of the experiment.
Features:

The remaining columns, excluding EIN, were used as features to enhance the accuracy of the model. Initially, the Name column was removed as a feature, but upon reintroduction and binning, the overall accuracy improved. Additionally, Organization was temporarily removed but later added back as it did not significantly benefit the model.
Variables to Remove:

The EIN column was removed from the input data due to its high variability. Initially, the Name column was also removed, but it was later reintroduced and binned to reduce variability, resulting in improved model performance.
Compiling, Training, and Evaluating the Model:

Neurons, Layers, and Activation Functions:

Each hidden layer consisted of two neurons, with ReLU activation functions used for the hidden layers and a sigmoid activation function for the output layer. These choices were made based on experimentation and their effectiveness in the given context.
Achievement of Target Model Performance:

The target model performance of achieving an accuracy greater than 75% was successfully attained, with the final model achieving 79% accuracy compared to the initial 72-73% in the first two attempts.
Steps to Increase Model Performance:

Several steps were taken to increase model performance, including adding a third hidden layer, reducing the number of epochs from 200 to 100, and reintroducing and binning the Name column to provide additional useful data for the model.
Summary:

Originally, the model involved removing the EIN and Name columns from the dataset, binning the CLASSIFICATION and APPLICATION_TYPE columns, and creating two hidden layers. By reintroducing and binning the Name column and adding a third hidden layer, the model's accuracy increased significantly. There are various ways to further improve model accuracy, such as adjusting the number of neurons and hidden layers or experimenting with different activation functions. The steps taken in this analysis were effective in achieving the desired accuracy, but further experimentation could yield even better results.



